### ISSUE TITLE
Support code completion of Flutter asset filenames in Image.asset() constructor (etc)

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#46542]
Originally raised in https://github.com/Dart-Code/Dart-Code/issues/3433 by @bdlukaa.

It would be nice if code completion understood when arguments are paths to Flutter assets and could provide completion for them (similar to how `imports` have filename completion):

<img width="522" alt="Screenshot 2021-07-01 at 16 53 22" src="https://user-images.githubusercontent.com/1078012/124153907-de9c6480-da8c-11eb-97b7-5f09ae7123e0.png">

Some possible ideas:

- An annotation on the `String` parameter to mark this argument as a path
- Code-gen to produce constants that would naturally show up in the completion list (this may work better if the type wasn't just `String`, although I suspect there would be less friction using `Strings`)

**Comment by bwilkerson** [comment]
The annotation would be a better solution because it would be available to all API authors. Would we need variations to control which platform the path was for?

**Comment by DanTup** [comment]
Based on https://github.com/flutter/flutter/issues/32039, I believe these should always use forward slashes (although it doesn't seem to be explicitly documented).

**Comment by bwilkerson** [comment]
Yes, I believe that Flutter uses POSIX-style paths for all of its APIs. But to make it generally useful for other package authors we might need more flexibility. Of course, one could argue that file paths that need to be platform-specific should be constructed using the `path` package's `join` function, but maybe there's an opportunity for a lint if we know that it's suppose to be a platform-specific path.

**Comment by jacob314** [comment]
Bumping the priority slightly as this is a quite frequently used API in Flutter so providing quality completions would have a significant impact.

**Comment by scheglov** [comment]
Do we really need generality here?
[This query](https://github.com/search?q=path%3A*.dart+flutter+%22Image.asset%28%27%22&type=code) shows 73k+ files that use `Image.asset()` with a string literal.
For [comparison](https://github.com/search?q=path%3A*.dart+flutter+%2FImage%5C.asset%5C%28%5Cw%2B%2F&type=code), using an identifier is less often, about 55k. Not zero, but I suspect that most of these do not originate from static literals that would benefit from a hypothetical `@assertUriString`.

So, I propose implementing support only for `Image.assert()`.

**Comment by bwilkerson** [comment]
Generality isn't an absolute requirement, but it is a very strong goal. We should always look for ways to make the support we're adding be something that has value for all of the package authors in the ecosystem rather than just for one or two hardcoded methods in the SDK libraries or Flutter packages. When package authors can express their intention for how an API is to be used, and we can use that information to provide a better dev experience, then we multiply the impact of our efforts.

**Comment by scheglov** [comment]
I opened https://github.com/flutter/flutter/issues/144937 with a proposal for Flutter to consider.

**Comment by scheglov** [comment]
It looks that Flutter folks are positive about these marker annotations.
So, I expect that they will accept the PR when we sent it.
But I wonder if I should first implement the corresponding DAS functionality, and see if it works well.
And add new API (annotations) to Flutter only then.
WDYT?

**Comment by bwilkerson** [comment]
I'm fine with whatever you're comfortable with.

Implementing the completion side (possibly by special casing on use case) would tell us whether the annotations are correctly defined. It shouldn't be too hard to update them if we need to, as long as they haven't shipped yet, but having some confidence in the design might be nice.
